src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -560,6 +560,14 @@ private BoundExpression BindSimpleBinaryOperator(BinaryExpressionSyntax node, Di
 
             if (hasErrors)
```

Comment:
>hasErrors [](start = 16, length = 9)

This is a strange ordering of the code. We've decided that there is an error, thus setting `hasErrors`. Then, we decide it isn't an error after all and ignore that flag.

I think the code should be organized this way:

``` c#
if (best.HasValue)
{
    ... code for successful binary operator overload resolution
}
else if (GetTupleCardinality(...) ...)
{
    ... code for tuple equality
}
else
{
    hasErrors = true;
    ... code for error case
}
```

Reply:
Unfortunately, it didn't clean up that way. The problem is that we could have `best.HasValue` but it turns out that was just the `object` equality which gets rejected on line 556:
`hasErrors = isObjectEquality && !BuiltInOperators.IsValidObjectEquality(Conversions, leftType, leftNull, rightType, rightNull, ref useSiteDiagnostics);`

Most tuple equality scenarios go through this path.

---
In reply to: [168056587](https://github.com/dotnet/roslyn/pull/23356#discussion_r168056587) [](ancestors = 168056587)